[RFC] vim-patch: 7.4.{389, 659} + small cleanup

Pyrohh opened the pull request:
based on #2429 and #2347

```
  Problem:    When 'ruler' is set the preferred column is reset. (Issue 339)
Solution:   Don't set curswant when redrawing the status lines.
```

  https://code.google.com/p/vim/source/detail?r=v7-4-659

  Original patch:

``` diff
  diff --git a/src/nvim/option.c b/src/nvim/option.c
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -7193,7 +7193,7 @@ did_set_string_option(opt_idx, varp, new
 #endif

     if (curwin->w_curswant != MAXCOL
-            && (options[opt_idx].flags & (P_CURSWANT | P_RCLR)) != 0)
+            && (options[opt_idx].flags & (P_CURSWANT | P_RALL)) != 0)
    curwin->w_set_curswant = TRUE;

 #ifdef FEAT_GUI
@@ -8203,7 +8203,7 @@ set_bool_option(opt_idx, varp, value, op

     comp_col();                /* in case 'ruler' or 'showcmd' changed */
     if (curwin->w_curswant != MAXCOL
-            && (options[opt_idx].flags & (P_CURSWANT | P_RCLR)) != 0)
+            && (options[opt_idx].flags & (P_CURSWANT | P_RALL)) != 0)
    curwin->w_set_curswant = TRUE;
     check_redraw(options[opt_idx].flags);

@@ -8738,7 +8738,7 @@ set_num_option(opt_idx, varp, value, err

     comp_col();                /* in case 'columns' or 'ls' changed */
     if (curwin->w_curswant != MAXCOL
-            && (options[opt_idx].flags & (P_CURSWANT | P_RCLR)) != 0)
+            && (options[opt_idx].flags & (P_CURSWANT | P_RALL)) != 0)
    curwin->w_set_curswant = TRUE;
     check_redraw(options[opt_idx].flags);

diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    659,
+/**/
     658,
 /**/
     657,
```

justinmk commented:
Our scripts expect no space in `vim-patch:7.4.389`. And we don't want the scripts to be flexible on this, because we don't want ambiguity.

Pyrohh commented:
@justinmk fixed, I had no idea that was the reason.

justinmk commented:
Nice, QB finished before travis :)

Pyrohh commented:
Ooh, that's pretty quick. Thanks for QB @jszakmeister.

Pyrohh commented:
Added patch from #2347

justinmk commented:
Test is failing on travis. I think 7.4.659 is N/A for Neovim: I cannot reproduce the issue as described in 

https://code.google.com/p/vim/issues/detail?id=339

It is a redraw issue, so it is quite possible that we don't need the changes.

Pyrohh commented:
@justinmk Sounds good, I'll look into that. I have finals soon, so I'll set this to WIP for now as I won't have time.

Pyrohh commented:
Well it seems I have more free time than expected, so I'll deal with this.

> Test is failing on travis. I think 7.4.659 is N/A for Neovim: I cannot reproduce the issue as described in
> 
> https://code.google.com/p/vim/issues/detail?id=339

Are you doing it with `nvim -u NONE`? I can reproduce it fine using the instructions provided. I'll write a test regardless.

Commit:
vim-patch:7.4.389

based on #2429 by @Hettomei

Problem:    Still sometimes Vim enters Replace mode when starting up.
Solution:   Use a different solution in detecting the termresponse and
            location response. (Hayaki Saito)

https://code.google.com/p/vim/source/detail?r=v7-4-389

The upstream patch touched a lot of files which don't exist in nvim,
hence the small size of this commit.

Commit:
src: Misc. cleanup

These macros (and global) haven't been used since
3baba1e7bc6698e6bc9f1d37fce88b30d6274bc9

Commit:
vim-patch:7.4.659 + test

Problem:    When 'ruler' is set the preferred column is reset. (Issue 339)
Solution:   Don't set curswant when redrawing the status lines.

https://code.google.com/p/vim/source/detail?r=v7-4-659

Helped-by: David Bürgin <676c7473@gmail.com>
Based on #2347 by @pvinis

Pyrohh commented:
I added a test and repushed. It's my busted first test ever, so It might not be as idiomatic as possible.

glts commented on test/functional/ui/screen_basic_spec.lua:
```diff
@@ -243,6 +241,30 @@ describe('Screen', function()
     end)
+
+    -- https://code.google.com/p/vim/issues/detail?id=339
+    it([[setting 'ruler' doesn't reset the preferred column]], function()
```
The `describe` block is labelled 'insert mode' but this test isn't about insert mode. Looking at the other `describe` blocks in this file perhaps moving the test case into a `describe('normal mode', ...` block of its own seems fair.

Pyrohh replied:
:+1:

glts commented:
Apart from the one comment, LGTM.

Pyrohh commented:
Thanks @glts, I'll merge after tests pass if no one objects.

Pyrohh commented:
@justinmk the infamous test 61 is failing on one machine, do you think this shouldn't be merged yet? I ask because @elmart made a comment 1-2 weeks ago about getting all tests passing before merging; not sure what everyone else thinks given how arbitrary the test 61 failures seem.

justinmk commented:
:+1: OK to merge.
